File: app/src/main/java/br/pedroso/upcomingmovies/repository/MoviesRepositoryImpl.java
package br.pedroso.upcomingmovies.repository;

import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import br.pedroso.upcomingmovies.domain.Movie;
import br.pedroso.upcomingmovies.domain.MoviesRepository;
import br.pedroso.upcomingmovies.network.entities.RetrofitMovieEntity;
import br.pedroso.upcomingmovies.network.services.TheMovieDbService;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class MoviesRepositoryImpl implements MoviesRepository {
    private final TheMovieDbService theMovieDbService;

    @Inject
    public MoviesRepositoryImpl(TheMovieDbService theMovieDbService) {
        this.theMovieDbService = theMovieDbService;
    }

    @Override
    public Single<List<Movie>> listUpcomingMovies() {
        return theMovieDbService.listUpcomingMovies()
                .map(retrofitMovieEntities -> retrofitMovieEntities.stream().map(new RetrofitMovieEntityToMovieMapper()).collect(Collectors.toList()))
                .subscribeOn(Schedulers.io());
    }

    @Override
    public Single<Movie> getMovieDetails(Integer movieId) {
        return theMovieDbService.getMovieDetails(movieId)
                .map(retrofitMovieEntity -> new RetrofitMovieEntityToMovieMapper().apply(retrofitMovieEntity))
                .subscribeOn(Schedulers.io());
    }

    @Override
    public Single<List<Movie>> listSimilarMovies(Integer movieId) {
        return theMovieDbService.listSimilarMovies(movieId)
                .map(retrofitMovieEntities -> retrofitMovieEntities.stream().map(new RetrofitMovieEntityToMovieMapper()).collect(Collectors.toList()))
                .subscribeOn(Schedulers.io());
    }

    private class RetrofitMovieEntityToMovieMapper implements java.util.function.Function<RetrofitMovieEntity, Movie> {
        private final String IMAGE_TMDB_BASE_URL = "https://image.tmdb.org/t/p/w500";

        private final Double MAX_VOTE_AVERAGE = 10.0;

        @Override
        public Movie apply(RetrofitMovieEntity retrofitMovieEntity) {
            return new Movie.Builder(retrofitMovieEntity.getId(), retrofitMovieEntity.getTitle())
                    .setPosterPath(IMAGE_TMDB_BASE_URL + retrofitMovieEntity.getPosterPath())
                    .setOverview(retrofitMovieEntity.getOverview())
                    .setReleaseDate(retrofitMovieEntity.getReleaseDate())
                    .setVoteAverage(retrofitMovieEntity.getVoteAverage() / MAX_VOTE_AVERAGE)
                    .build();
        }
    }
}

File: app/src/main/kotlin/br/pedroso/upcomingmovies/repository/MoviesRepositoryImpl.kt
package br.pedroso.upcomingmovies.repository

import br.pedroso.upcomingmovies.domain.Movie
import br.pedroso.upcomingmovies.domain.MovieDetails
import br.pedroso.upcomingmovies.domain.MoviesRepository
import br.pedroso.upcomingmovies.network.services.TheMovieDbService
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import javax.inject.Inject

class MoviesRepositoryImpl @Inject constructor(
    private val theMovieDbService: TheMovieDbService,
    private val retrofitMovieEntityToMovieMapper: RetrofitMovieEntityToMovieMapper,
) : MoviesRepository {
    override suspend fun listUpcomingMovies(): List<Movie> {
        return theMovieDbService.listUpcomingMovies().results.map(retrofitMovieEntityToMovieMapper)
    }

    override suspend fun getMovieDetails(movieId: Int): MovieDetails = coroutineScope {
        val movie = async { getMovie(movieId) }

        val similarMovies = async { listSimilarMovies(movieId) }

        MovieDetails(movie.await(), similarMovies.await())
    }

    private suspend fun getMovie(movieId: Int): Movie {
        return retrofitMovieEntityToMovieMapper(theMovieDbService.getMovieDetails(movieId))
    }

    private suspend fun listSimilarMovies(movieId: Int): List<Movie> {
        return theMovieDbService.listSimilarMovies(movieId).results.map(
            retrofitMovieEntityToMovieMapper
        )
    }
}
